From ef9e0a1863db79316a69bb1f150bbbe0f10f7f47 Mon Sep 17 00:00:00 2001
From: MarshallZhan-wiwynn <marshall_zhan@wiwynn.com>
Date: Wed, 12 Feb 2025 16:13:31 +0800
Subject: [PATCH 1013/1016] leds: support CAT9532 led controller

Summary:
Description
- CAT9532 used the opposite polarity in register that sets on/off.
- Compare the input and the led select register before set led state.

Motivation
- support CAT9532 led controller

Test Plan:
- Build code : pass
- Check LED status : pass

lore link:
https://lore.kernel.org/all/20250212071525.1148988-1-marshall_zhan@wiwynn.com/
---
 drivers/leds/leds-pca955x.c | 37 +++++++++++++++++++++++++++++++++----
 1 file changed, 33 insertions(+), 4 deletions(-)

diff --git a/drivers/leds/leds-pca955x.c b/drivers/leds/leds-pca955x.c
index 240a31e813e7..20fd9e38a891 100644
--- a/drivers/leds/leds-pca955x.c
+++ b/drivers/leds/leds-pca955x.c
@@ -13,6 +13,7 @@
  *	PCA9550		2-bit driver		0x60 .. 0x61
  *	PCA9551		8-bit driver		0x60 .. 0x67
  *	PCA9552		16-bit driver		0x60 .. 0x67
+ *	CAT9532		16-bit driver		0x60 .. 0x67
  *	PCA9553/01	4-bit driver		0x62
  *	PCA9553/02	4-bit driver		0x63
  *
@@ -276,6 +277,20 @@ static int pca955x_read_psc(struct pca955x *pca955x, int n, u8 *val)
 	return 0;
 }
 
+static int pca955x_read_input_bit(struct pca955x *pca955x, int led_num)
+{
+        u8 cmd = led_num / 8;
+        int ret;
+
+        ret = i2c_smbus_read_byte_data(pca955x->client, cmd);
+        if (ret < 0) {
+                dev_err(&pca955x->client->dev, "%s: reg 0x%x, err %d\n", __func__, led_num, ret);
+                return ret;
+        }
+        return (ret >> (led_num % 8)) & 1;
+
+}
+
 static enum led_brightness pca955x_led_get(struct led_classdev *led_cdev)
 {
 	struct pca955x_led *pca955x_led = led_to_pca955x(led_cdev);
@@ -288,12 +303,15 @@ static enum led_brightness pca955x_led_get(struct led_classdev *led_cdev)
 		return ret;
 
 	switch (pca955x_ledstate(ls, pca955x_led->led_num % 4)) {
-	case PCA955X_LS_LED_ON:
 	case PCA955X_LS_BLINK0:
 		ret = LED_FULL;
 		break;
+	case PCA955X_LS_LED_ON:
 	case PCA955X_LS_LED_OFF:
-		ret = LED_OFF;
+		if (pca955x_read_input_bit(pca955x, pca955x_led->led_num))
+			ret = LED_FULL;
+		else
+			ret = LED_OFF;
 		break;
 	case PCA955X_LS_BLINK1:
 		ret = pca955x_read_pwm(pca955x, 1, &pwm);
@@ -314,6 +332,8 @@ static int pca955x_led_set(struct led_classdev *led_cdev,
 	int reg = pca955x_led->led_num / 4;
 	int bit = pca955x_led->led_num % 4;
 	u8 ls;
+	u8 ls_last_state;
+	int input_bit;
 	int ret;
 
 	mutex_lock(&pca955x->lock);
@@ -322,6 +342,7 @@ static int pca955x_led_set(struct led_classdev *led_cdev,
 	if (ret)
 		goto out;
 
+	ls_last_state = pca955x_ledstate(ls, bit);
 	if (test_bit(pca955x_led->led_num, &pca955x->active_blink)) {
 		if (value == LED_OFF) {
 			clear_bit(pca955x_led->led_num, &pca955x->active_blink);
@@ -333,10 +354,18 @@ static int pca955x_led_set(struct led_classdev *led_cdev,
 	} else {
 		switch (value) {
 		case LED_FULL:
-			ls = pca955x_ledsel(ls, bit, PCA955X_LS_LED_ON);
+			input_bit = pca955x_read_input_bit(pca955x, pca955x_led->led_num);
+			if (ls_last_state == input_bit)
+				ls = pca955x_ledsel(ls, bit, PCA955X_LS_LED_ON);
+			else
+				ls = pca955x_ledsel(ls, bit, PCA955X_LS_LED_OFF);
 			break;
 		case LED_OFF:
-			ls = pca955x_ledsel(ls, bit, PCA955X_LS_LED_OFF);
+			input_bit = pca955x_read_input_bit(pca955x, pca955x_led->led_num);
+			if (ls_last_state == input_bit)
+				ls = pca955x_ledsel(ls, bit, PCA955X_LS_LED_OFF);
+			else
+				ls = pca955x_ledsel(ls, bit, PCA955X_LS_LED_ON);
 			break;
 		default:
 			/*
-- 
2.47.3

